core/imds/src/test/java/software/amazon/awssdk/imds/Ec2MetadataClientTest.java

```diff
@@ -0,0 +1,391 @@
+            .hasMessageContaining("401")
+            .isInstanceOf(SdkClientException.class);
+
+        WireMock.verify(1, putRequestedFor(urlPathEqualTo(TOKEN_RESOURCE_PATH))
+            .withHeader(EC2_METADATA_TOKEN_TTL_HEADER, equalTo("21600")));
+        WireMock.verify(1, getRequestedFor(urlPathEqualTo(AMI_ID_RESOURCE))
+            .withHeader(TOKEN_HEADER, equalTo("some-token")));
+    }
+
+    @Test
+    public void getToken_doesNotRetryOn403() {
+
+        stubFor(put(urlPathEqualTo(TOKEN_RESOURCE_PATH)).willReturn(aResponse().withStatus(403)));
+        stubFor(get(urlPathEqualTo(AMI_ID_RESOURCE)).willReturn(aResponse().withBody("{}")));
+
+        assertThatThrownBy(() -> ec2MetadataClient.get("/latest/meta-data/ami-id"))
+            .hasMessageContaining("token")
+            .isInstanceOf(SdkClientException.class);
+        WireMock.verify(exactly(1), putRequestedFor(urlPathEqualTo(TOKEN_RESOURCE_PATH))
+            .withHeader(EC2_METADATA_TOKEN_TTL_HEADER, equalTo("21600")));
+        WireMock.verify(0, getRequestedFor(urlPathEqualTo(AMI_ID_RESOURCE))
+            .withHeader(TOKEN_HEADER, equalTo("some-token")));
+    }
+
+    @Test
+    public void getToken_doesNotRetryOn401() {
+
+        stubFor(put(urlPathEqualTo(TOKEN_RESOURCE_PATH)).willReturn(aResponse().withStatus(401)));
+        stubFor(get(urlPathEqualTo(AMI_ID_RESOURCE)).willReturn(aResponse().withBody("{}")));
+
+        assertThatThrownBy(() -> ec2MetadataClient.get("/latest/meta-data/ami-id"))
+            .hasMessageContaining("token")
+            .isInstanceOf(SdkClientException.class);
+        WireMock.verify(exactly(1), putRequestedFor(urlPathEqualTo(TOKEN_RESOURCE_PATH))
+            .withHeader(EC2_METADATA_TOKEN_TTL_HEADER, equalTo("21600")));
+        WireMock.verify(0, getRequestedFor(urlPathEqualTo(AMI_ID_RESOURCE))
+            .withHeader(TOKEN_HEADER, equalTo("some-token")));
+    }
+
+    @Test
+    public void get_failsThriceWithFixedDelay() {
+
+        stubFor(put(urlPathEqualTo(TOKEN_RESOURCE_PATH)).willReturn(aResponse().withBody("some-token")));
+        stubFor(get(urlPathEqualTo(AMI_ID_RESOURCE)).willReturn(aResponse().withFixedDelay(2_000)));
+
+        SdkHttpClient client = UrlConnectionHttpClient.builder().socketTimeout(Duration.ofMillis(500)).build();
+        Ec2MetadataClient ec2MetadataRequest = Ec2MetadataClient.builder()
+                                                                .httpClient(client)
+                                                                .endpoint(URI.create("http://localhost:" + mockMetadataEndpoint.port()))
+                                                                .build();
+        assertThatThrownBy(() -> ec2MetadataRequest.get("/latest/meta-data/ami-id"))
+            .hasMessageContaining("Exceeded maximum number of retries.")
+            .isInstanceOf(SdkClientException.class);
+
+        WireMock.verify(4, putRequestedFor(urlPathEqualTo(TOKEN_RESOURCE_PATH))
```

Comment:
Can we extract `4` to a static final instance variable to make it clear it refers to the default allowed attempts?